src/hevm/src/EVM/SMT.hs

```diff
@@ -687,6 +732,146 @@ checkSat (SolverGroup taskQueue) script = do
+parseW256 = parseSC
+
+parseInteger :: SpecConstant -> Integer
+parseInteger = parseSC
+
+parseW8 :: SpecConstant -> Word8
+parseW8 = parseSC
+
+parseErr :: (Show a) => a -> b
+parseErr res = error $ "Internal Error: cannot parse solver response: " <> show res
+
+parseVar :: TS.Text -> Expr EWord
+parseVar = Var
+
+parseBlockCtx :: TS.Text -> Expr EWord
+parseBlockCtx "origin" = Origin
+parseBlockCtx "coinbase" = Coinbase
+parseBlockCtx "timestamp" = Timestamp
+parseBlockCtx "blocknumber" = BlockNumber
+parseBlockCtx "difficulty" = Difficulty
+parseBlockCtx "gaslimit" = GasLimit
+parseBlockCtx "chainid" = ChainId
+parseBlockCtx "basefee" = BaseFee
+parseBlockCtx t = error $ "Internal Error: cannot parse " <> (TS.unpack t) <> " into an Expr"
+
+parseFrameCtx :: TS.Text -> Expr EWord
+parseFrameCtx name = case TS.unpack name of
+  ('c':'a':'l':'l':'v':'a':'l':'u':'e':'_':frame) -> CallValue (read frame)
+  ('c':'a':'l':'l':'e':'r':'_':frame) -> Caller (read frame)
+  ('a':'d':'d':'r':'e':'s':'s':'_':frame) -> Address (read frame)
+  t -> error $ "Internal Error: cannot parse " <> t <> " into an Expr"
+
+getVars :: (TS.Text -> Expr EWord) -> SolverInstance -> [TS.Text] -> IO (Map (Expr EWord) W256)
+getVars parseFn inst names = Map.mapKeys parseFn <$> foldM getVar mempty names
+  where
+    getVar :: Map TS.Text W256 -> TS.Text -> IO (Map TS.Text W256)
+    getVar acc name = do
+      raw <- getValue inst (T.fromStrict name)
+      let
+        parsed = case parseCommentFreeFileMsg getValueRes (T.toStrict raw) of
+          Right (ResSpecific (valParsed :| [])) -> valParsed
+          r -> parseErr r
+        val = case parsed of
+          (TermQualIdentifier (
+            Unqualified (IdSymbol symbol)),
+            TermSpecConstant sc)
+              -> if symbol == name
+                 then parseW256 sc
+                 else error "Internal Error: solver did not return model for requested value"
+          r -> parseErr r
+      pure $ Map.insert name val acc
+
+getBufs :: SolverInstance -> [TS.Text] -> IO (Map (Expr Buf) ByteString)
+getBufs inst names = foldM getBuf mempty names
+  where
+    getLength :: TS.Text -> IO Int
+    getLength name = do
+      val <- getValue inst (T.fromStrict name <> "_length")
+      len <- case parseCommentFreeFileMsg getValueRes (T.toStrict val) of
+        Right (ResSpecific (parsed :| [])) -> case parsed of
+          (TermQualIdentifier (Unqualified (IdSymbol symbol)), (TermSpecConstant sc))
+            -> if symbol == (name <> "_length")
+               then pure $ parseW256 sc
+               else error "Internal Error: solver did not return model for requested value"
+          res -> parseErr res
+        res -> parseErr res
+      pure $ if len <= num (maxBound :: Int)
+             then fromIntegral len
+             else error $ "Internal Error: buffer: "
+                       <> (TS.unpack name)
+                       <> " is too large to be represented in a ByteString. Length: "
```

Comment:
I can still hit this case, I guess maybe then we need to implement some counterexample shrinking? perhaps we can implement some kind of binary search where we constrain calldata to be smaller and check if the property can still be violated?

Reply:
I think the solver just gives back a length that it knows will work. Basically, there is likely some bound on it, but the solver is not asked to give a minimal value that satisfies that bound, only to give a value that satisfies it, which may be a very large number. This can only be overcome with a query that gives some optimization bound, say, `min(length)`, potentially with some timeout. Then the solver will either give a minimal value, or timeout and give the "best so far". Z3 should be able to do this, but I don't know if CVC5 supports optimization criteria. We could get away with a short bound, as it should be able to quickly make it relatively small (<2**20). However, it would actually be a full-on query, not just asking the solver for the value of a variable.